Sources/OpenAPIRuntime/Interface/Body.swift

```diff
@@ -0,0 +1,535 @@
+// MARK: - Creating the Body
+
+extension Body {
+
+    public convenience init(
+        data: DataType,
+        length: Length
+    ) {
+        self.init(
+            dataChunks: [data],
+            length: length
+        )
+    }
+
+    public convenience init(
+        data: DataType
+    ) {
+        self.init(
+            dataChunks: [data],
+            length: .known(data.count)
+        )
+    }
+
+    public convenience init<S: Sequence>(
+        dataChunks: S,
+        length: Length,
+        iterationBehavior: IterationBehavior
+    ) where S.Element == DataType {
+        self.init(
+            sequence: .init(WrappedSyncSequence(sequence: dataChunks)),
+            length: length,
+            iterationBehavior: iterationBehavior
+        )
+    }
+
+    public convenience init<C: Collection>(
+        dataChunks: C,
+        length: Length
+    ) where C.Element == DataType {
+        self.init(
+            sequence: .init(WrappedSyncSequence(sequence: dataChunks)),
+            length: length,
+            iterationBehavior: .multiple
+        )
+    }
+
+    public convenience init<C: Collection>(
+        dataChunks: C
+    ) where C.Element == DataType {
+        self.init(
+            sequence: .init(WrappedSyncSequence(sequence: dataChunks)),
+            length: .known(dataChunks.map(\.count).reduce(0, +)),
+            iterationBehavior: .multiple
+        )
+    }
+
+    public convenience init(
+        stream: AsyncThrowingStream<DataType, any Error>,
+        length: Body.Length
+    ) {
+        self.init(
+            sequence: .init(stream),
+            length: length,
+            iterationBehavior: .single
+        )
+    }
+
+    public convenience init(
+        stream: AsyncStream<DataType>,
+        length: Body.Length
+    ) {
+        self.init(
+            sequence: .init(stream),
+            length: length,
+            iterationBehavior: .single
+        )
+    }
+
+    public convenience init<S: AsyncSequence>(
+        sequence: S,
+        length: Body.Length,
+        iterationBehavior: IterationBehavior
+    ) where S.Element == DataType {
+        self.init(
+            sequence: .init(sequence),
+            length: length,
+            iterationBehavior: iterationBehavior
+        )
+    }
+}
+
+// MARK: - Consuming the body
+
+extension Body: AsyncSequence {
+    public typealias Element = DataType
+    public typealias AsyncIterator = Iterator
+    public func makeAsyncIterator() -> AsyncIterator {
+        if iterationBehavior == .single {
+            lock.lock()
+            defer {
+                lock.unlock()
+            }
+            guard !locked_iteratorCreated else {
+                fatalError(
+                    "OpenAPIRuntime.Body attempted to create a second iterator, but the underlying sequence is only safe to be iterated once."
+                )
+            }
+            locked_iteratorCreated = true
+        }
+        return sequence.makeAsyncIterator()
```

Comment:
We shouldn't have to do this at all. If we just delegate to call `makeAsyncIterator` on the underlying sequence then that base async sequence should enforce if it is `unicast` or `multicast`.

Reply:
That's a fair request, and something we also considered. However I find that the way that crash happens is pretty opaque and doesn't really help the user understand what went wrong, and how to fix it. We're just catching this scenario early and emitting a more descriptive error, which I think is generally a good approach? WDYT?